New Base Template

Mark-H opened the pull request:
### Why is it needed

When installing a MODX setup, there is always an empty base template that shows nothing to the screen - just a white page. This is confusing, and not very helpful to people who are just getting started with MODX as it may look like something went wrong. This is a missed opportunity. 
### What does it do

This pull request introduces a new Base Template. It's self-contained (except the Open Sans font being loaded from a CDN, however that is loaded in a progressively enhanced way, so no blocking), so there are no additional assets or "themes" introduced here. It also doesn't rely on any extras, nor does it add additional resources. This is simply making use of a currently missed opportunity. 

This new base template gives the user a virtual high five for getting MODX installed, and provides pointers to a number of resources. First and foremost, there's a big button to go the manager to start editing and building the site. There are various links to the official MODX RTFM both for the top level Revo docs, and specific subjects (e.g. friendly urls, the different types of elements). There's a high level max-two-sentences-per-subject introduction of how MODX works. There are links to the MODX forums, and also modx.org for the MODX Community chat there. 

Finally, at the bottom there are logos of the different places that people can get extras from (MODX, modmore, modstore, extras.io). As Extras are critical to MODX sites these give people a head start to figuring out where to find them, and I can imagine that for people who are already familiar with them might find it useful as quick click through to those websites to see what's new, or to get their api keys for the private ones. The logos are shuffled with some javascript to prevent playing favourites. 

The design is neutral, it doesn't force a "MODX look" in the same way half the web looks like wordpress. By implementing the content of the page as actual content in the automatically created Home page (vs hardcoded in the template or using lexicon tags), it gives the first time user an accessible way to start editing a page to see what happens. 

**This is not a starter theme**, nor does it replace the need for that, which other people in the community are working on. Starter themes give people a more thorough insight in structuring and building websites, as well as a starting point for tweaking a site that would actually go live. See #12859.

[Here's what it looks like on a desktop](https://cloud.githubusercontent.com/assets/312944/12356522/78538824-bba3-11e5-917e-aba3f7140870.png), [and here's a view of it on mobile](https://cloud.githubusercontent.com/assets/312944/12356602/c9fb81ea-bba3-11e5-86f4-d8f97251abce.png).
### Related issue(s)/PR(s)

None that I could find, though this idea has been discussed on Slack recently. 
### Credits

@sytskehaagsma did great work on the design/front-end 
Ivo Pakvis helped edit the copy I wrote
Then I just threw it all together into this PR

jpdevries commented:
+1

sottwell commented:
+1

markwillis82 commented:
Looks good. Well done all :) :+1:

Omeryl commented:
Erm, does this mean that the community project now officially sanctions modstore and moremore? Eh .. that's a little weird to me @Mark-H

Omeryl commented:
Also, the copyright date says 2015.

sottwell commented:
Maybe somebody should submit a feature request for a built-in "now" MODX placeholder to output the current date. It could be modified using the "date" output modifier to display just the year if desired.

JimKillock commented:
You know, "now" isn't completely correct for a copyright notice. Copyright fixes by the creation year. Design elements would be copyright 2015, if that is when @Mark-H designed them. Content would be copyright by the year of publication. Often websites incorrectly assert the current year as the year of publication, even when they also display prior published-on dates, which are a better indication of the correct copyright year.

Mark-H commented:
> Erm, does this mean that the community project now officially sanctions modstore and moremore? Eh .. that's a little weird to me @Mark-H

This is the reasoning for including the logos: 

> Finally, at the bottom there are logos of the different places that people can get extras from. As Extras are critical to MODX sites these give people a head start to figuring out where to find them, and I can imagine that for people who are already familiar with them might find it useful as quick click through to those websites to see what's new, or to get their api keys for the private ones.

It definitely needs some label or heading saying "Here's where you can get Extras", right now it looks more like sponsor logos than it probably should. 

I think it only makes sense for the project to embrace things like modmore, modstore and extras.io. These are companies and communities that add tremendous value to the project/product. Of course I'm a businessman and want to see my company its logo everywhere, but I really do believe that in terms of getting people started after an install, that they should be introduced in a way to these other vendors one way or another. It could also be another list in the sidebar "Where to get Extras" with text links.

Omeryl commented:
@Mark-H I'm definitely only stating my personal opinion here, but I'd be a bit happier if it had a proper header and maybe a disclaimer. 

(not speaking on behalf of MODX LLC on this one).

Mark-H commented:
I'll try to post some quick alternative mockups for the extras vendors later today, open to ideas from others too ;)

adamwintle commented:
This is a great idea, and I can't believe nobody has thought of this before. A total missed opportunity! 

I would not use the Google CDN because it is blocked in China and other regions, maybe just use Arial or another system font. (MODX is of course global :) I'd also suggest not hot-linking the logo from modx.com, the user may be behind a firewall, offline or, for some reason modx.com might be unaccessible. Maybe use a base64 image of the MODX logo to reduce on files so this welcome page is a one-page drop-in. 

Perhaps in the "Get Help" sidebar of useful links there could be a link to the Ambassadors and Professionals for users who want to hire browse those listings. 

The "Lear more about MODX" sidebar could also include a link to [Discover MODX](http://discovermodx.com/) website, as well as the official MODX Twitter and Facebook pages so beginners can join the community immediately in any way that suits them.

Maybe the "There are a different type of resources" section is a big too much, perhaps the overall text could be reduced to keep it short and sweet.

adamwintle commented on setup/templates/base_template.tpl:
```diff
@@ -0,0 +1,262 @@
+        </ul>
+    </div>
+</div>
+<div class="disclaimer">
```
This could use HTML5 `<footer>` tag

sottwell commented:
Does it need to go into any detail about the different types of resources? My thought it that it should be as simple and basic as possible, just enough to get somebody started.

adamwintle commented:
For the copyright, it could say "2005-present", I've seen some websites do this and I like the concept. Depending on where you live the copyright laws are based off of when the works creation date, meaning you only need to state the year it was made.

adamwintle commented:
For the row of logos, in my opinion its fine to include third-party logos, some of the best plugins can be found at the third-party sites. Perhaps this could have a heading to make it clear what the logos represent, e.g. "Get started by downloading MODX Extras from these sites"

jaygilmore commented:
I am not a fan of the logos. I think all, including MODX Extras could be in a nice list in the sidebar like other resources we're referring people to. The challenge I have with this document is it does require review regularly to maintain it.

Jako commented:
Nice work, much better than nothing (take this serious). 

The logos are maybe a bit too much. What if other third-party sites occur? Who decides which logo is shown and which not? That would be my main issue with that.

beau-gosse commented:
+1

jpdevries commented:
> Erm, does this mean that the community project now officially sanctions modstore and moremore? Eh .. that's a little weird to me @Mark-H

To sanction is to:
 > give official permission or approval for (an action)

@Omeryl What is the community project's __official__ process for giving permission to or approving third party package providers? AFAIK there is no official process, meaning it isn't possible to sanction a provider in the first place. So no, nothing is being sanctioned. Yes, we are asking for permission and approval to inform new users where they can find Extras. With logos. :scream: 

There is a link to buy commercial support directly from the LLC on the Help page. Does that mean the _community project_ approves or prioritizes LLC consults over others?

![](http://j4p.us/2q2L29343p0j/Screen%20Shot%202016-01-15%20at%209.13.08%20PM.png)

No, I don't think it does. Contextually relevant information, such as where to find Emergency Help or where to find Extras, is presented to users where it would be most relevant to them. If a user is in an emergency and seeking immediate help, having the Emergency Support option to contact the LLC is helpful to them. It should be there. By the same logic, a MODX newbie may appreciate a few links to where they can find freemium and premium Extras outside of the MODX provider.

People can pull MODX Extras from multiple providers and we are informing first-time users of this newfound ability.

> The challenge I have with this document is it does require review regularly to maintain it.

Is anyone up for taking the responsibility for maintaining the document and/or seeking maintainers to keep it up to date? Do we have any reason to believe the collective community wouldn't utilize Pull Requests to keep this doc up to date? I've heard non-coders speak of ways they would like to contribute. This seems like a perfect opportunity for that.

> What if other third-party sites occur? Who decides which logo is shown and which not? 

Shouldn't that be the same people who decide on any other changes to the product, the MODX Integrators? When said third-party providers arise they are free to submit a PR that adds their logo. In said PR the matter at hand could be discussed, voted on, etc. This is similar to the  process projects like cdnjs and Font-Awesome use.

bezumkin commented:
@Mark-H Great idea! +1
@jpdevries Totally agree.

sottwell commented:
I kind of liked the logos myself. I don't think using logos for links to something third-party is necessarily any kind of official sanction, simply an acknowledgement. Consider the "Professionals Directory". Although I do think, design-wise, that they are a little in-you-face large. But that's more a matter of personal taste, I suppose.

Jako commented:
> Shouldn't that be the same people who decide on any other changes to the product, the MODX Integrators?

Maybe that has the same transparency issue. Who are the integrators, how did they get this role etc. A fast google search for 'modx integrators' does not show a clear answer on this.

jpdevries commented:
@Jako There are 6 integrators and they are part of this group
https://github.com/orgs/modxcms/teams/integration-managers-revolution

That's all I know

sottwell commented:
Apparently we plebeians are not privileged to view that page.

pixelchutes commented:
I personally think this is a great idea. +1

It is a throwback to the original base template / sample content from the Evo days, an undoubtedly will serve newcomers learning the framework by helping them connect the dots.

As a user's experience grows, it may become less "valuable" over time (e.g. future instances, etc), but at the end of the day I see it as benefiting the entire MODX ecosystem as a whole, which is a good thing. :wink:

Mark-H commented:
> The logos are maybe a bit too much. What if other third-party sites occur? Who decides which logo is shown and which not? That would be my main issue with that.

That could be a case-by-case determination.. personally the reason for these 4 are:
- they've all been around for some time 
- they provide a number of different extras, there are some other parties that sell a single extra (Bitego, BobsGuides) but if including those might mean there will be dozens of logos by next year. In that case, perhaps a page in the RTFM might be better suited to contain a full list of all potential sources/vendors. 

I can imagine a new site sending a pull request might get declined on their launch day, but might get included six months later. If there's no consensus among integrators, the community could vote on it or something.

Mark-H commented:
> I am not a fan of the logos. I think all, including MODX Extras could be in a nice list in the sidebar like other resources we're referring people to. The challenge I have with this document is it does require review regularly to maintain it.

I don't see too much change being necessary to this base template, but if it is there are more people able of proposing a change to the tpl files (everyone with a github account that can hit the "edit" button), than there are people that have access to edit the RTFM :P

sottwell commented:
Yes, a link to a RTFM page that maintains a known list of sources for extras would be one way of dealing with it.

sottwell commented:
Meanwhile, while we quibble over minutiae, the question remains about having some kind of reasonable default content upon installation. Anything would be better than what we have now, which is virtually nothing.

Mark-H commented:
> Maybe that has the same transparency issue. Who are the integrators, how did they get this role etc. A fast google search for 'modx integrators' does not show a clear answer on this.

> Apparently we plebeians are not privileged to view that page.

<img width="230" alt="screen shot 2016-01-16 at 19 38 06" src="https://cloud.githubusercontent.com/assets/312944/12373838/dc139922-bc88-11e5-914f-846c8f4d4b7e.png">

To my knowledge, all integrators were asked to become one by @opengeek. I don't know if there is any process for it in the LLC, or if it's at his discretion. There is a private slack channel for integrators that is (pretty rarely) used to discuss issues that need to be discussed, and I recall there at some point being a chat about possible future integrators. 

To become an integrator, people should probably just show them to be active contributors. There's a lot people can do without requiring access to the green merge button, including sending PRs, triaging issues, reviewing PRs etc.

Mark-H commented:
Back to this pull request :D Great feedback from all, thanks. 

How's this, with a header to properly indicate why the logos are following?

<img width="1083" alt="screen shot 2016-01-16 at 20 11 09" src="https://cloud.githubusercontent.com/assets/312944/12373964/50f55a60-bc8d-11e5-929d-6ca050fd23eb.png">

Also changed the modmore logo to be blue on white rather than white on blue to make a bit calmer, and [asked Extras.io](https://twitter.com/mark_hamstra/status/688417585503797248) if they also have a reversed colour one to use instead. Based on feedback here I've also tweaked the text about resources (mostly just removed the list and added a link to the RTFM instead. Based on feedback from @christianseel I've slightly reduced the logo size to allow more whitespace between them. 

(Commit to follow shortly)

opengeek commented:
I have no problem with there being a directory of 3rd party extras providers. But I do have a problem with product releases being tied directly to a version of this portion of the content. Is there some way we can externalize this portion of the content and deliver it dynamically in an efficient and cacheable way without affecting the integrity of this "base template"?

jpdevries commented:
@opengeek what if we use Grunt to automate a reg ex for updating the dynamic bits?

Mark-H commented:
@opengeek I'm not sure if I understand you correctly, but if you mean the "2.5.0-dev" bit, that comes from the settings_version system setting and isn't hardcoded into the template.

opengeek commented:
No, I'm asking if we deliver the footer containing the logos to the template from a hosted source so it is not a part of the deployment itself?

jpdevries commented:
Oh I see. hm, like host it on S3 or some API and load it over XHR?

To keep the document simple, might be best to do with VanillaJS (like I did with the FOUT loading of the font). @Mark-H let me know if you want me to wip that up.

sottwell commented:
I'm inclined to think that we are getting rather far afield here. The whole point is to have a simple, basic "welcome to MODX" page on installation. An initial trial installation may very well be on a localhost platform. Having parts of that page dependent on external sources is kind of defeating that purpose.

jpdevries commented:
> An initial trial installation may very well be on a localhost platform. Having parts of that page dependent on external sources is kind of defeating that purpose.

This is a good point. I still think we could cater to this progressively by progressively enhancing a simple external link to a "MODX Package Providers" page into the dynamic content of logos for the current providers.

```html
<footer id="providers" data-xhr-endpoint="http://api.modx.com/providers/all"><!-- use JS to hit this endpoint and overwrite innerHTML with response -->
  <p>You can extend MODX by installing Extras from these <a href="http://modx.com/providers">MODX&nbsp;Providers</a>.</p>
</footer>
```

If the aren't connected to the internet, the XHR request will fail but but they will still be able to access and bookmark a link.

adamwintle commented:
I completely agree with @sottwell, nothing on this page should be external and everything should be self-contained. Ideally assets required for the page should be base64 if possible so the /assets/ folder is totally clean on a fresh install. 

We don't know the users setup or network, and to presume they are online would be wrong. I have worked on new sites when offline before, or from tethering from a mobile connection; it would be a bad user experience for them to get broken images links on the welcome page. 

I'd also not over-complicate things by injecting sections of the page from third-party sources; we would then need to detect weather the user is connected to the internet so we have a graceful fallback.

As for the rows of Extra provider logos, I say we just go with it for now and see if it produces problems in the future. It can always easily be removed if it causes an issue. 

@Mark-H Can you post an updated screenshot from the latest updates?

Mark-H commented:
Been chatting on Slack with Jason to understand his concerns about wanting to load the data from elsewhere (primarily so changes to the list don't rely on the release process of MODX to get updated in installations, but there were a few others) and I'm personally not sold. That would be a very rare scenario (vendors don't just drop off the map, especially the premium ones, and there are regular core updates with bug fixes), but it does add complexity and needs to be made manageable (and accountable) elsewhere. 

There is still a need for a full starter/sample site, which is well-built using a variety of extras, TVs, settings and what not. This is not that, nor does it aspire to be, and the delivery mechanism of that should indeed be kept separate from the MODX core repo. Probably as a transport package installed when the user ticks a checkbox in the set up. I'd be happy to dedicate some time to such an effort as well, but to me that's a completely different project and not affected by this PR.

Mark-H commented:
Latest full page screenshot:

![base template 2015-01-16](https://cloud.githubusercontent.com/assets/312944/12374429/471b6332-bc9b-11e5-8a0f-ddf9eede27f7.png)

Also includes the fix by @jpdevries to load the font asynchronously, so that part is also more China and offline friendly now. MODX logo is also loaded locally from the manager folder.

adamwintle commented:
@Mark-H How about adding the official [Twitter](https://twitter.com/modx) and [Facebook](https://www.facebook.com/modxcms) social links in the Get Help sidebar under the Slack link? Some newcomers might not want to join the Slack chat room and feel more comfortable engaging via Twitter or Facebook. 

I'll ask my friend in China to see if he can test to async font loading, but I think it should do the trick.

jpdevries commented:
> That would be a very rare scenario (vendors don't just drop off the map, especially the premium ones, and there are regular core updates with bug fixes), but it does add complexity and needs to be made manageable (and accountable) elsewhere.

Agreed. Also, I think the same logic of "what if one is added or removed" applies to the "Learn more about MODX" resources section. I think that would update more than the providers list. I'm not sure why the packages list in particular is a concern. 

I think users understand the information is up to date as of when the release was tagged.

For the copy, if people are still concerned with if it feels like the project "sanctions" these providers maybe the headline could read "You Can Extend MODX with Extras". More informative, less suggestive. 

> I'd also not over-complicate things by injecting sections of the page from third-party sources; we would then need to detect weather the user is connected to the internet so we have a graceful fallback.

Or in other words assume we aren't online (HTML first) and try to reach out for an update but fail gracefully. Example in https://github.com/modxcms/revolution/pull/12855#issuecomment-172252631
I agree it complicates things though. Would be nice to have it all upfront and simple.

sottwell commented:
If I recall correctly, MODX LLC already has a cert, it just needs to be applied.

adamwintle commented:
Yes, always assume the user isn't online.  @jpdevries if you guys want to try injecting the row of logos via some JS then go for it; but my vote would be to just leave out the JS and have the row of logos as normal HTML and images.

jpdevries commented:
> Yes, always assume the user isn't online. @jpdevries if you guys want to try injecting the row of logos via some JS then go for it; but my vote would be to just leave out the JS and have the row of logos as normal HTML and images.

I don't really want to. More so implying I'd be willing to if it is deemed necessary for this to be accepted. 

With this PR https://github.com/Mark-H/revolution/pull/4 no more typographic orphans on the page 

![](http://j4p.us/2J0Z2B3q241v/Screen%20Shot%202016-01-16%20at%201.18.16%20PM.png) :x:

mindeffects commented:
A startpage is a good first step. +1
The logos, well, maybe not at the bottom but in the aside column? 
Demo content (like we had in Evo) could be an option during a fresh setup. It could include some basic getResources and wayfinder calls, too. Would make the first steps easier for noobs.

The "setup" could be a "download script" first, which checks, if Revo advanced can be used (else fallback to trad), downloads the stuff, extract the zip and starts the real "setup". This would speed up the first MODX steps for at least 90% of the users. But that's another story/fr/pr. ;-)

Mark-H committed:
Update Extras.io logo to be color-on-white like the rest

Mark-H committed:
Merge remote-tracking branch 'jpdevries/new-base-template-orphans' into new-base-template

Mark-H committed:
Fix some nbsp's

Mark-H committed:
Shuffle vendor logos with javascript, while leaving MODX in first place

Mark-H committed:
Add twitter and facebook links

Mark-H commented:
Couple of further tweaks, including preventing orphaned words (thanks @jpdevries) and an updated Extras.io logo. I've also added some javascript that shuffles the order of the different vendors (except the official MODX repo, that's always first) to make sure there's no preference to any particular vendor. Also added links to twitter and facebook (@adamwintle). 

Any outstanding issues to address or would this be ready for merge, 48 comments later? :P 

[Latest screenshot here](https://cloud.githubusercontent.com/assets/312944/12393468/17821740-bdf6-11e5-9e93-493c42b9f3a2.png)

adamwintle commented:
Looking great now! :) @Mark-H do we need the word "Official" in "Official MODX Forums" and "Official MODX Blog"?

Mark-H commented:
I thought that would be best given there's also community sources in the list, to give those particular links more credibility.

OptimusCrime commented:
Wow, awesome! +1 for this.

For people wanting to use CDN, remember offline people!

Mark-H commented:
There's no reliance on CDN anymore (updated the PR description); there's just a nicer font being loaded asynchronously as a progressive enhancement after the page is rendered. Fully self-contained beyond that :)

OptimusCrime commented:
Ah, sorry @Mark-H. I did not check the diffs that carefully and I did not read all the conversations.

Again, nice work! :D

Mark-H commented:
:tada:

sebastian-marinescu commented:
:fireworks:

Ruslan-Aleev commented:
@Mark-H  you can post a request in the base template that the users MODX CMS placed on modx.com link on their websites to increase the popularity MODX CMS.
